src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, VecDeque};
+
+use bytes::Bytes;
+use risingwave_common::catalog::TableId;
+use risingwave_common::hash::VirtualNode;
+use risingwave_hummock_sdk::key::FullKey;
+use risingwave_hummock_sdk::table_watermark::{ReadTableWatermark, WatermarkDirection};
+
+use crate::hummock::iterator::{Forward, HummockIterator};
+use crate::hummock::value::HummockValue;
+use crate::hummock::HummockResult;
+use crate::monitor::StoreLocalStatistic;
+
+pub struct SkipWatermarkIterator<I> {
+    inner: I,
+    watermarks: BTreeMap<TableId, ReadTableWatermark>,
+    remain_watermarks: VecDeque<(TableId, VirtualNode, WatermarkDirection, Bytes)>,
+}
+
+impl<I: HummockIterator<Direction = Forward>> SkipWatermarkIterator<I> {
+    pub fn new(inner: I, watermarks: BTreeMap<TableId, ReadTableWatermark>) -> Self {
+        Self {
+            inner,
+            remain_watermarks: VecDeque::new(),
+            watermarks,
+        }
+    }
+
+    fn reset_watermark(&mut self) {
+        self.remain_watermarks = self
+            .watermarks
+            .iter()
+            .flat_map(|(table_id, read_watermarks)| {
+                read_watermarks
+                    .vnode_watermarks
+                    .iter()
+                    .map(|(vnode, watermarks)| {
+                        (
+                            *table_id,
+                            *vnode,
+                            read_watermarks.direction,
+                            watermarks.clone(),
+                        )
+                    })
+            })
+            .collect();
+    }
+
+    /// Advance watermark until no watermark remains or the first watermark can possibly
+    /// filter out the current or future key.
+    ///
+    /// Return a flag indicating whether the current key will be filtered by the current watermark.
+    fn advance_watermark(&mut self) -> bool {
+        if self.inner.is_valid() {
+            let key = self.inner.key();
+            let key_table_id = key.user_key.table_id;
+            let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+            while let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front()
```

Comment:
> https://github.com/risingwavelabs/risingwave/pull/13429#discussion_r1405830051

Not only because of this. It's like merging two sorted lists, and we only check and pop the item at the front.